Fix extra incorrectly shown on installation error

sepehr-rs opened the pull request:
Fixes #13618
This resolves the issue, though I’m not certain it’s the ideal place for the fix. If needed, we can move the solution further down in the stack.

sepehr-rs committed:
Remove extras from pip error message when installing packages

sepehr-rs committed:
Add news entry

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Update 13618.bugfix.rst

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Move removal of extra down the stack

sepehr-rs committed:
Fix linting issue

sepehr-rs committed:
Merge branch 'main' into fix-extras

notatallshaw reviewed (changes requested):
Update the test `test_install_special_extra` in `tests/functional/test_install_extras.py` to include something like this at the end:

```python
    assert "No matching distribution found for missing_pkg" in result.stderr, str(result)
    assert "extra ==" not in result.stderr, str(result)
```

Then I think you're going to find two situations you're not handeling right now:

1. When the requirement has a parent, you need to move your string manipulation to the top of `_report_single_requirement_conflict` so it also gets applied to the `req_disp` variable for logging
2. The logging and raising of raise DistributionNotFound in `PackagingFinder.find_requirement` in `src/pip/_internal/index/package_finder.py`

notatallshaw commented:
I'm realizing I wrote the above comment in a bit of a rush and I've not verified the output is actually misleading in that test. So feel free to correct me if I made a mistake, I'll take a other look in the next few days when I have time to review again.

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Update tests and coverage

danielhollas reviewed (commented):
Not a pip developer, but hopefully some helpful review comments...

Also this needs more tests I think, right now it seems that only one of the two modified paths is tested?

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Update code to address reviews

sepehr-rs commented:
Hi @danielhollas, thanks a lot for the helpful review! I've addressed your comments, but I’m not entirely clear on your note about the assertion test. Could you clarify what you meant?

danielhollas reviewed (commented):
@sepehr-rs thanks. You can disregard my test assertion comment, but I think there's still an issue with how you override `req_disp` variable.

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Update factory.py

sepehr-rs commented:
Hi, sorry for the delay. Things have been hectic over here in Iran and I only recently got access to GitHub again. I’ve committed the changes proposed by @danielhollas. Please let me know what you think!

sepehr-rs committed:
Merge branch 'main' into fix-extras

notatallshaw commented:
Sorry, I think this PR still needs some more review and work, but I'm very unlikely to have time to review before 26.1.

danielhollas commented on src/pip/_internal/resolution/resolvelib/factory.py:
```diff
@@ -693,7 +694,7 @@ def _report_single_requirement_conflict(
         else:
             req_disp = f"{req} (from {parent.name})"
-
+        req_disp = re.sub(r';\s*extra\s*==\s*".*?"', "", str(req)).strip()
```
I believe there's still a bug here, since you're overwriting the `req_disp` defined above. I think you want to do this:


```suggestion
        req_disp = re.sub(r';\s*extra\s*==\s*".*?"', "", str(req_disp)).strip()
```

sepehr-rs replied:
Thanks so much for your kind words!
I did this intentionally to minimize pip’s output, but I totally understand your point and it makes sense.

danielhollas replied:
> I did this intentionally to minimize pip’s output,

I am not sure what you mean?

sepehr-rs replied:
With `req_disp`:
```
ERROR: Could not find a version that satisfies the requirement definitely-does-not-exist-xyz123 (from dummy-pkg[optimizer]) (from versions: none)
ERROR: No matching distribution found for definitely-does-not-exist-xyz123 (from dummy-pkg[optimizer])
```

With `req`:
```
ERROR: Could not find a version that satisfies the requirement definitely-does-not-exist-xyz123 (from versions: none)
ERROR: No matching distribution found for definitely-does-not-exist-xyz123
```

danielhollas replied:
I see. But I think the `from XXX` part can be important so should be kept (and also removing it would go beyond what was discussed in the linked issue). Thanks for making the change.

danielhollas commented on src/pip/_internal/index/package_finder.py:
```diff
@@ -1016,16 +1016,19 @@ def _format_versions(cand_iter: Iterable[InstallationCandidate]) -> str:
+            # pip sometimes includes environment markers like `;extra == "foo"`
+            # in requirement strings, which is confusing in error messages.
+            # Strip them so the user sees only the actual missing requirement.
+            req_str = re.sub(r';\s*extra\s*==\s*".*?"', "", str(req)).strip()
```
Let's make this consistent with the other file

```suggestion
            req_disp = re.sub(r';\s*extra\s*==\s*".*?"', "", str(req)).strip()
```

danielhollas reviewed (commented):
> Things have been hectic over here in Iran and I only recently got access to GitHub again.

Oof, what an understatement, I am very sorry. :-( 

Thank you for the changes,  I think that there is at least one thing that needs fixing, otherwise looks reasonable to me, but I am not a pip maintainer, just trying to help with a review :-)

sepehr-rs committed:
Update src/pip/_internal/resolution/resolvelib/factory.py

Co-authored-by: Daniel Hollas <danekhollas@gmail.com>

sepehr-rs committed:
Update package_finder.py

sepehr-rs committed:
Merge branch 'main' into fix-extras

danielhollas commented on src/pip/_internal/resolution/resolvelib/factory.py:
```diff
@@ -693,7 +694,7 @@ def _report_single_requirement_conflict(
         else:
             req_disp = f"{req} (from {parent.name})"
-
+        req_disp = re.sub(r';\s*extra\s*==\s*".*?"', "", str(req_disp)).strip()
```
nit: `req_disp` is already a string so no need for the cast (sorry, should have caught this earlier)

```suggestion
        req_disp = re.sub(r';\s*extra\s*==\s*".*?"', "", req_disp).strip()
```

sepehr-rs replied:
Nice catch, thanks a lot!

sepehr-rs committed:
Update src/pip/_internal/resolution/resolvelib/factory.py

Co-authored-by: Daniel Hollas <danekhollas@gmail.com>

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs commented:
Gentle ping. I believe I've addressed all review comments and follow-up feedback. I think this may be ready for another review. I'd appreciate any feedback when someone has time.

danielhollas reviewed (commented):
Looks good to me, but not a pip maintainer. :-)

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Merge branch 'main' into fix-extras

sepehr-rs committed:
Merge branch 'main' into fix-extras

ichard26 commented:
@notatallshaw I'll leave the review to you. I'm not a fan of using regex substitution, but on a quick glance, it doesn't seem like the packaging marker API allows us to filter out extra markers easily...

ichard26 commented:
I'm deferring this to pip 26.3.
